### PR TITLE
Feature/elliott/reduce ranks

### PR DESCRIPTION
**Pull request by nselliott**
This enables SPIO reads when running on a smaller number of processors.  It has a restriction that the original files must have been produced in a file-per-processor mode.  Removing this restriction will require some further development.

### CODE REVIEW

**Comment by nselliott** [comment]
Domain-overloading is still ultimately a task for the calling code.  This can get them the data they need for it, but it's up to them to determine what they do with it.

File: src/axom/sidre/docs/sphinx/parallel_io_concepts.rst [hunk 1]
```diff
@@ -46,17 +46,29 @@ management (such as using burst buffers if available).
...
+
+If ``read()`` is called using "sidre_hdf5" to read data that was created on
+a larger number of processors, this will work only in the case that the data
+was written in a file-per-processor mode (M ranks to M files).  In this case
+the data in the Group being filled with file input will look a bit different
+than in other usage patterns, since a Group on one rank will end up with data
+from multiple ranks.  An integer scalar View named ``reduced_input_ranks``
+will be added to the Group with the value being the number of ranks that
+wrote the files.  The data from each output rank will be read into subgroups
+located at ``rank_*******/sidre_input`` in the input Group's data hierarchy.
```
Comment (kennyweiss) [inline src/axom/sidre/docs/sphinx/parallel_io_concepts.rst]: Is ``*******`` a stand-in for a number?  If so, would something like ``{%07d}`` be clearer?

**Comment by cyrush** [comment]
Ok reading more not I am a bit confused. This changes the partitioning and the resulting trees in a strange way.

Can we white board this a bit (as a larger team)?

File: src/axom/sidre/docs/sphinx/parallel_io_concepts.rst (hunk 1 above)
Comment (nselliott) [reply src/axom/sidre/docs/sphinx/parallel_io_concepts.rst]: Yes, that would be better.

**Comment by nselliott** [comment]
> Ok reading more not I am a bit confused. This changes the partitioning and the resulting trees in a strange way.
> 
> Can we white board this a bit (as a larger team)?

I did bring this up for discussion in a meeting a few weeks ago.  In this usage a single processor is getting data that was produced by multiply outputting processors, and that output may consist of identically-structured Sidre hierarchies, so there needed to be an extra layer added to keep the data from each processor separate.  The "reduced_input_ranks" View is an indicator that something different has happened and the user needs to be aware.

File: src/axom/sidre/tests/spio/spio_parallel.cpp [hunk 2]
```diff
@@ -809,6 +810,145 @@ TEST(spio_parallel, parallel_increase_procs)
...
+
+  int top_input_rank = 1;
+  if (num_ranks <= 2)
+  {
+    top_input_rank = 0;
+  }
+
+  // Split the communicator so that ranks up to and including
+  // top_input_rank have their own communicator for the input step.
+  MPI_Comm split_comm;
```
Comment (gzagaris) [inline src/axom/sidre/tests/spio/spio_parallel.cpp]: I think you need to call `MPI_Comm_free( split_comm )` at the end to free that communicator and properly cleanup any MPI state.

**Comment by cyrush** [comment]
@nselliott sorry I wasn't there, I do understand the challenge with unique names. I am wondering how this will compose with domain overloaded case that we will need to support. Seems like we should clarify the use cases.

`reduced_input_ranks` is documented, which is great but it doesn't seem very intuitive to me.

**Comment by nselliott** [comment]
@cyrush I suppose I see domain-overloading as tangential to the scope of this feature addition.  Someone can create a domain-overloaded mesh without ever using this feature, as long as they run from start to finish without changing the size of their MPI allocation.  IOManager itself doesn't know anything about domains--it does I/O of Groups and Views that have data that could be anything.

**Comment by cyrush** [comment]
@nselliott sounds like you have a specific use case in mind for this --  maybe we can talk about it.

Seems to low bandwidth here, maybe at a future meeting?

**Comment by nselliott** [comment]
The use case is that a restart dump exists and the user wants to re-launch the run on a smaller processor allocation.  In order for this to work, the application code needs to be programmed to know what to do with data that once existed on the higher ranks that no longer exist.  It could handle this with a domain overloading approach, or with a repartition and communication of data.  IOManager has no knowledge of what the calling code wants to do with the data--all that it's doing here is getting the data from the files into memory.

**Pull request merged**